Pair reportExtraMemoryAllocated with reportExtraMemoryVisited for SourceBuffer, collection index caches and ImageData

Ahmad-S792 opened the pull request:
#### aa27d10f9deaaa9ca8c5e4aec0cfc1cfd4af05eb
<pre>
Pair reportExtraMemoryAllocated with reportExtraMemoryVisited for SourceBuffer, collection index caches and ImageData
<a href="https://bugs.webkit.org/show_bug.cgi?id=296676">https://bugs.webkit.org/show_bug.cgi?id=296676</a>
<a href="https://rdar.apple.com/157080340">rdar://157080340</a>

Reviewed by Yusuke Suzuki.

`deprecatedReportExtraMemory` and `reportExtraMemoryAllocated` are not
interchangeable. `deprecatedReportExtraMemory` adds the cost to
`m_deprecatedExtraMemorySize`, which keeps contributing to `extraMemorySize()`
until the next full collection zeroes it. `reportExtraMemoryAllocated` adds
nothing to that total by itself -- it only does `didAllocate()` plus
`collectIfNecessaryOrDefer()` (and re-reports through the barrier if the owning
cell is already black). The cost only enters `extraMemorySize()` via the matching
`reportExtraMemoryVisited()` performed while the owning cell is visited.

So swapping one for the other without a visit counterpart hides the memory from
the collector for good, which is what the comment on
`Heap::reportExtraMemoryAllocated` warns about: &quot;Be sure to call both of these
functions: Calling only one may trigger catastropic memory growth.&quot;

This patch therefore only converts the call sites whose owning wrapper already
reports the same cost while being visited. `SourceBuffer`, `HTMLCollection` and
`NodeList` are annotated `[ReportExtraMemoryCost]`, so CodeGeneratorJS.pm emits
`visitor.reportExtraMemoryVisited(thisObject-&gt;wrapped().memoryCost())` into their
`visitChildren()`, and `memoryCost()` already covers exactly what these call
sites report. They were double counting until now: once into the deprecated
bucket at allocation time, and again into `m_extraMemorySize` on every
collection. `ImageData` adopts `[ReportExtraMemoryCost]` here so its pixel data
is paired the same way.

* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::reportExtraMemoryAllocated):
`SourceBuffer.idl` is `[ReportExtraMemoryCost]` and `memoryCost()` returns
`sizeof(SourceBuffer) + m_extraMemoryCost`, which is the total this function has
just recorded, so reporting the delta as allocated is correctly paired.

* Source/WebCore/dom/CollectionIndexCache.cpp:
(WebCore::reportExtraMemoryAllocatedForCollectionIndexCache):
Every `CollectionIndexCache` owner is a `CachedHTMLCollection`, `ChildNodeList`
or `LiveNodeList`, and both `HTMLCollection.idl` and `NodeList.idl` are
`[ReportExtraMemoryCost]` with a `memoryCost()` that returns the index cache and
named element cache costs. The capacity deltas reported here are already covered
on the visit side.

* Source/WebCore/html/ImageData.idl:
Annotate the interface `[ReportExtraMemoryCost]` so `JSImageData` reports the
pixel data in `finishCreation()` and re-reports it in `visitChildren()`.

* Source/WebCore/html/ImageData.h:
(WebCore::ImageData::memoryCost const):
* Source/WebCore/html/ImageData.cpp:
(WebCore::ImageData::ImageData):
Add `memoryCost()`. The generated `visitChildren()` calls it from a GC thread, so
the byte length is computed once in the constructor instead of being read off the
array buffer view on demand: `ArrayBufferView::byteLength()` consults
`isDetached()` and, for a resizable or growable-shared buffer, dereferences
`m_buffer`, neither of which is safe to race with a detach. `m_data` is assigned
only in the constructor, so a `const size_t` snapshot is accurate for the
lifetime of the object.

* Source/WebCore/bindings/js/JSImageDataCustom.cpp:
(WebCore::toJSNewlyCreated):
Drop the manual report now that the generated wrapper does it, and note where the
reporting moved to.

Canonical link: <a href="https://commits.webkit.org/319569@main">https://commits.webkit.org/319569@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0ed859b46ab33fc78649cd80d72905bd1886729e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/179093 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/53032 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/46827 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/188922 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/143402 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b88ef044-7f50-4787-877a-92d737570d34) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/180956 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/54325 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/52742 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/139662 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/96755 "2 failures") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/0c5cd293-aadf-4cf8-8876-459ae5095d43) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/182050 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/43259 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/160419 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/120108 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/24b3e8ab-934b-4a59-a52c-22fcc4a0d900) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/41930 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/40273 "Passed tests") | [⏳ 🧪 jsc-wpe](https://ews-build.webkit.org/#/builders/JSC-Tests-WPE-EWS "Waiting to run tests") | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/152330 "Build is in progress. Recent messages:OS: Tahoe (26.5), Xcode: 26.5; Checked out pull request; run-api-tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/39922 "Passed tests") | [✅ 🛠 gtk3-gcc](https://ews-build.webkit.org/#/builders/284/builds/229 "Built successfully") | | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/45638 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/44518 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/191142 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/52702 "Built successfully") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/148895 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/40540 "Built successfully and passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/53382 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/36548 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/148646 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/43329 "Passed tests") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/125653 "Built successfully") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/120467 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/51900 "Built successfully") | [❌ 🧪 mac-site-isolation](https://ews-build.webkit.org/#/builders/185/builds/14901 "") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/51285 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/51526 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/51346 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

Ahmad-S792 committed:
Reduce `deprecatedReportExtraMemory` usage throughout WebKit
https://bugs.webkit.org/show_bug.cgi?id=296676
rdar://157080340

Reviewed by NOBODY (OOPS!).

This patch fixes old FIXME to adopt `reportExtraMemoryVisited` by
replacing `deprecatedReportExtraMemory` with `reportExtraMemoryAllocated`.

* Source/JavaScriptCore/API/JSBase.cpp:
(JSReportExtraMemoryCost):
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::reportExtraMemoryAllocated):
* Source/WebCore/bindings/js/JSImageDataCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/dom/CollectionIndexCache.cpp:
(WebCore::reportExtraMemoryAllocatedForCollectionIndexCache):
* Source/WebCore/html/HTMLImageLoader.cpp:
(WebCore::HTMLImageLoader::notifyFinished):
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::pauseAfterDetachedTask):
* Source/WebCore/svg/graphics/SVGImage.cpp:
(WebCore::SVGImage::reportApproximateMemoryCost const):

Constellation reviewed (commented):
When using `reportExtraMemoryAllocated`, you also need to use `reportExtraMemoryVisited` appropriately.

Ahmad-S792 committed:
Pair reportExtraMemoryAllocated with reportExtraMemoryVisited for SourceBuffer, collection index caches and ImageData
https://bugs.webkit.org/show_bug.cgi?id=296676
rdar://157080340

Reviewed by NOBODY (OOPS!).

`deprecatedReportExtraMemory` and `reportExtraMemoryAllocated` are not
interchangeable. `deprecatedReportExtraMemory` adds the cost to
`m_deprecatedExtraMemorySize`, which keeps contributing to `extraMemorySize()`
until the next full collection zeroes it. `reportExtraMemoryAllocated` adds
nothing to that total by itself -- it only does `didAllocate()` plus
`collectIfNecessaryOrDefer()` (and re-reports through the barrier if the owning
cell is already black). The cost only enters `extraMemorySize()` via the matching
`reportExtraMemoryVisited()` performed while the owning cell is visited.

So swapping one for the other without a visit counterpart hides the memory from
the collector for good, which is what the comment on
`Heap::reportExtraMemoryAllocated` warns about: "Be sure to call both of these
functions: Calling only one may trigger catastropic memory growth."

This patch therefore only converts the call sites whose owning wrapper already
reports the same cost while being visited. `SourceBuffer`, `HTMLCollection` and
`NodeList` are annotated `[ReportExtraMemoryCost]`, so CodeGeneratorJS.pm emits
`visitor.reportExtraMemoryVisited(thisObject->wrapped().memoryCost())` into their
`visitChildren()`, and `memoryCost()` already covers exactly what these call
sites report. They were double counting until now: once into the deprecated
bucket at allocation time, and again into `m_extraMemorySize` on every
collection. `ImageData` adopts `[ReportExtraMemoryCost]` here so its pixel data
is paired the same way.

* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::reportExtraMemoryAllocated):
`SourceBuffer.idl` is `[ReportExtraMemoryCost]` and `memoryCost()` returns
`sizeof(SourceBuffer) + m_extraMemoryCost`, which is the total this function has
just recorded, so reporting the delta as allocated is correctly paired.

* Source/WebCore/dom/CollectionIndexCache.cpp:
(WebCore::reportExtraMemoryAllocatedForCollectionIndexCache):
Every `CollectionIndexCache` owner is a `CachedHTMLCollection`, `ChildNodeList`
or `LiveNodeList`, and both `HTMLCollection.idl` and `NodeList.idl` are
`[ReportExtraMemoryCost]` with a `memoryCost()` that returns the index cache and
named element cache costs. The capacity deltas reported here are already covered
on the visit side.

* Source/WebCore/html/ImageData.idl:
Annotate the interface `[ReportExtraMemoryCost]` so `JSImageData` reports the
pixel data in `finishCreation()` and re-reports it in `visitChildren()`.

* Source/WebCore/html/ImageData.h:
(WebCore::ImageData::memoryCost const):
* Source/WebCore/html/ImageData.cpp:
(WebCore::ImageData::ImageData):
Add `memoryCost()`. The generated `visitChildren()` calls it from a GC thread, so
the byte length is computed once in the constructor instead of being read off the
array buffer view on demand: `ArrayBufferView::byteLength()` consults
`isDetached()` and, for a resizable or growable-shared buffer, dereferences
`m_buffer`, neither of which is safe to race with a detach. `m_data` is assigned
only in the constructor, so a `const size_t` snapshot is accurate for the
lifetime of the object.

* Source/WebCore/bindings/js/JSImageDataCustom.cpp:
(WebCore::toJSNewlyCreated):
Drop the manual report now that the generated wrapper does it, and note where the
reporting moved to.

Ahmad-S792 committed:
Pair reportExtraMemoryAllocated with reportExtraMemoryVisited for SourceBuffer, collection index caches and ImageData
https://bugs.webkit.org/show_bug.cgi?id=296676
rdar://157080340

Reviewed by Yusuke Suzuki.

`deprecatedReportExtraMemory` and `reportExtraMemoryAllocated` are not
interchangeable. `deprecatedReportExtraMemory` adds the cost to
`m_deprecatedExtraMemorySize`, which keeps contributing to `extraMemorySize()`
until the next full collection zeroes it. `reportExtraMemoryAllocated` adds
nothing to that total by itself -- it only does `didAllocate()` plus
`collectIfNecessaryOrDefer()` (and re-reports through the barrier if the owning
cell is already black). The cost only enters `extraMemorySize()` via the matching
`reportExtraMemoryVisited()` performed while the owning cell is visited.

So swapping one for the other without a visit counterpart hides the memory from
the collector for good, which is what the comment on
`Heap::reportExtraMemoryAllocated` warns about: "Be sure to call both of these
functions: Calling only one may trigger catastropic memory growth."

This patch therefore only converts the call sites whose owning wrapper already
reports the same cost while being visited. `SourceBuffer`, `HTMLCollection` and
`NodeList` are annotated `[ReportExtraMemoryCost]`, so CodeGeneratorJS.pm emits
`visitor.reportExtraMemoryVisited(thisObject->wrapped().memoryCost())` into their
`visitChildren()`, and `memoryCost()` already covers exactly what these call
sites report. They were double counting until now: once into the deprecated
bucket at allocation time, and again into `m_extraMemorySize` on every
collection. `ImageData` adopts `[ReportExtraMemoryCost]` here so its pixel data
is paired the same way.

* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::reportExtraMemoryAllocated):
`SourceBuffer.idl` is `[ReportExtraMemoryCost]` and `memoryCost()` returns
`sizeof(SourceBuffer) + m_extraMemoryCost`, which is the total this function has
just recorded, so reporting the delta as allocated is correctly paired.

* Source/WebCore/dom/CollectionIndexCache.cpp:
(WebCore::reportExtraMemoryAllocatedForCollectionIndexCache):
Every `CollectionIndexCache` owner is a `CachedHTMLCollection`, `ChildNodeList`
or `LiveNodeList`, and both `HTMLCollection.idl` and `NodeList.idl` are
`[ReportExtraMemoryCost]` with a `memoryCost()` that returns the index cache and
named element cache costs. The capacity deltas reported here are already covered
on the visit side.

* Source/WebCore/html/ImageData.idl:
Annotate the interface `[ReportExtraMemoryCost]` so `JSImageData` reports the
pixel data in `finishCreation()` and re-reports it in `visitChildren()`.

* Source/WebCore/html/ImageData.h:
(WebCore::ImageData::memoryCost const):
* Source/WebCore/html/ImageData.cpp:
(WebCore::ImageData::ImageData):
Add `memoryCost()`. The generated `visitChildren()` calls it from a GC thread, so
the byte length is computed once in the constructor instead of being read off the
array buffer view on demand: `ArrayBufferView::byteLength()` consults
`isDetached()` and, for a resizable or growable-shared buffer, dereferences
`m_buffer`, neither of which is safe to race with a detach. `m_data` is assigned
only in the constructor, so a `const size_t` snapshot is accurate for the
lifetime of the object.

* Source/WebCore/bindings/js/JSImageDataCustom.cpp:
(WebCore::toJSNewlyCreated):
Drop the manual report now that the generated wrapper does it, and note where the
reporting moved to.

Canonical link: https://commits.webkit.org/319569@main
